java-checks/src/main/java/org/sonar/java/checks/HashMapComparableKeyCheck.java

```diff
@@ -0,0 +1,147 @@
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeArguments;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+@Rule(key = "S3552")
+public class HashMapComparableKeyCheck extends BaseTreeVisitor implements JavaFileScanner, JavaVersionAwareVisitor {
+
+  private static final String JAVA_UTIL_MAP = "java.util.Map";
+  private static final String JAVA_UTIL_HASH_MAP = "java.util.HashMap";
+  private JavaFileScannerContext context;
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava8Compatible();
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    Type variableType = tree.type().symbolType();
+    if (variableType.isSubtypeOf(JAVA_UTIL_MAP)) {
+      ExpressionTree initializer = tree.initializer();
+      if (initializer != null) {
+        checkNewMap(initializer, nonComparableKeyTypeName(variableType));
+      }
+    } else {
+      super.visitVariable(tree);
+    }
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    Type variableType = tree.variable().symbolType();
+    if (variableType.isSubtypeOf(JAVA_UTIL_MAP)) {
+      checkNewMap(tree.expression(), nonComparableKeyTypeName(variableType));
+    } else {
+      super.visitAssignmentExpression(tree);
+    }
+  }
+
+  @Override
+  public void visitNewClass(NewClassTree tree) {
+    if (tree.symbolType().isSubtypeOf(JAVA_UTIL_HASH_MAP)) {
+      checkNewMap(tree, nonComparableKeyTypeName(tree.identifier().symbolType()));
+    } else {
+      super.visitNewClass(tree);
+    }
+  }
+
+  @CheckForNull
+  private static String nonComparableKeyTypeName(Type symbolType) {
+    Type type = symbolType;
+    if (type instanceof MethodJavaType) {
+      type = ((MethodJavaType) type).resultType();
+    }
+    if (type instanceof ParametrizedTypeJavaType) {
+      ParametrizedTypeJavaType parametrizedTypeJavaType = (ParametrizedTypeJavaType) type;
+      JavaType keyType = parametrizedTypeJavaType.substitution(parametrizedTypeJavaType.typeParameters().get(0));
+      if (!keyType.isTagged(JavaType.TYPEVAR) && !keyType.isSubtypeOf("java.lang.Comparable")) {
```

Comment:
I wonder we should not skip interfaces as well. To me, issues when interfaces are used as key are still legit, as we need to be sure that `Comparable` is implemented. So the correct solution would be that the interface which is used should also force implementation of `Comparable`. I would appreciate another point of view.